jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -1051,10 +1096,17 @@ else if (_counts.getHi() == Integer.MIN_VALUE)
                             long idleTimeout = getIdleTimeout();
                             if (idleTimeout > 0 && getThreads() > _minThreads)
                             {
-                                long last = _lastShrink.get();
+                                long last;
+                                long siNanos;
                                 long now = NanoTime.now();
-                                if (NanoTime.millisElapsed(last, now) > idleTimeout && _lastShrink.compareAndSet(last, now))
+                                long itNanos = TimeUnit.MILLISECONDS.toNanos(idleTimeout);
+                                if (NanoTime.elapsed(idleBaseline, now) > itNanos &&
+                                    NanoTime.elapsed(last = _lastShrink.get(), now) > (siNanos = getShrinkInterval()) &&
+                                    _lastShrink.compareAndSet(last, Math.max(last, now - itNanos) + siNanos))
```

Comment:
I'm not a big fan of the inline assignment as it is hard to read.
Also I'm not sure we actually need to check the idle timeout here, nor that the we need an idleBaseline.
Couldn't this just be:
```suggestion
                                if (NanoTime.elapsed(last, now) > getShrinkInterval() && _lastShrink.compareAndSet(last, now))
```

Reply:
b082e965aa986aa8ca1e83db84206c599dbeeec1 removes the inline variable assignment (and adds some comments for clarity).

iiuc though, we definitely do still want to check idle timeout here (and need `idleBaseline`), otherwise `idleTimeout` is effectively a ceiling on the amount of time a thread _might_ wait before considering itself potentially eligible to be removed from the pool, and I had thought we'd want `idleTimeout` to be a floor for the amount of time a thread _must_ be idle before considering itself to be eligible for removal.

If we don't check `idleTimeout` here, then the only place we're really using `idleTimeout` is in `idleJobPoll()` -- so a thread will always try to remove itself (wrt only the `shrinkInterval`), unless a job is immediately available. Respecting `idleTimeout` as a floor for determining when a thread is eligible to die should help the pool grow more consistently (to accommodate high load), while allowing (via idleTimeoutDecay) the pool to shrink more efficiently once we're past the `idleTimeout` window of a transient spike.